base/compiler/ssair/passes.jl

```diff
@@ -261,61 +406,118 @@ function getfield_elim_pass!(ir::IRCode, domtree)
         isa(field, QuoteNode) && (field = field.value)
         isa(field, Union{Int, Symbol}) || continue
 
-        intermediaries = IdSet()
-        phi_locs = Tuple{Int, Int}[]
-        def = walk_to_def(compact, stmt.args[2], intermediaries, is_getfield, phi_locs)
-        def !== nothing || continue
-        (def, defidx) = def
+        struct_typ = unwrap_unionall(widenconst(compact_exprtype(compact, stmt.args[2])))
+        (isa(struct_typ, DataType) && !struct_typ.mutable) || continue
 
-        if !is_getfield
-            (defidx == 0) && continue
-            mid, defuse = get!(defuses, defidx, (IdSet{Int}(), SSADefUse()))
+        def, typeconstraint = stmt.args[2], struct_typ
+
+        if struct_typ.mutable
+            isa(def, SSAValue) || continue
+            intermediaries = IdSet()
+            def = simple_walk(compact, def, (pi, idx)->push!(intermediaries, idx))
+            # Mutable stuff here
+            isa(def, SSAValue) || continue
+            mid, defuse = get!(defuses, def.id, (IdSet{Int}(), SSADefUse()))
             push!(defuse.defs, idx)
             union!(mid, intermediaries)
             continue
         end
-        # Step 3: Check if the definition we eventually end up at is either
-        # a tuple(...) call or Expr(:new) and perform replacement.
-        if is_tuple_call(ir, def) && isa(field, Int) && 1 <= field < length(def.args)
-            forwarded = def.args[1+field]
-        elseif isexpr(def, :new)
-            typ = def.typ
-            if isa(typ, UnionAll)
-                typ = unwrap_unionall(typ)
+
+        if isa(def, Union{OldSSAValue, SSAValue})
+            def, typeconstraint = simple_walk_constraint(compact, def, typeconstraint)
+        end
+
+        visited_phinodes = Any[]
+        if isa(def, Union{OldSSAValue, SSAValue, NewSSAValue}) && isa(compact[def], PhiNode)
+            leaves = walk_to_defs(compact, def, typeconstraint, visited_phinodes)
+        else
+            leaves = [def]
+        end
+
+        isempty(leaves) && continue
+
+        field = try_compute_fieldidx(struct_typ, stmt)
+        field === nothing && continue
+
+        r = lift_leaves(compact, stmt, result_t, field, leaves)
+        r === nothing && continue
+        lifted_leaves, any_undef = r
+
+        reverse_mapping = IdDict{Any, Any}(ssa => id for (id, ssa) in enumerate(visited_phinodes))
+
+        if any_undef
+            result_t = make_MaybeUndef(result_t)
+        end
+
+        # Insert PhiNodes
+        lifted_phis = map(visited_phinodes) do item
```

Comment:
since the front-end doesn't have domtree analysis, wrap this in a `let` block to explicitly capture variables (to avoid preventing inference from seeing the types of these variables everywhere in the containing function)